VDA-XP-Reward/contracts/VDAXPReward.sol

```diff
@@ -0,0 +1,277 @@
+        conversionRate = newRate;
+        emit UpdateConversionRate(orgVal, newRate);
+    }
+
+    /**
+     * @notice Validate the time of proof
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address. Used to revert
+     * @param curTime - Current time
+     * @param info - Claim information
+     */
+    function _validateProofTime(address didAddress, DateTime._DateTime memory curTime, ClaimInfo calldata info) internal virtual view {
+        if (curTime.year != info.issueYear || curTime.month != (info.issueMonth + 1)) {
+            revert InvalidProofTime(didAddress, info.typeId, info.issueYear, info.issueMonth);
+        } else if (curTime.month == 1) {
+            if (curTime.year != (info.issueYear + 1) || info.issueMonth != 12) {
+                revert InvalidProofTime(didAddress, info.typeId, info.issueYear, info.issueMonth);
+            }
+        }
+    }
+
+    /**
+     * @notice Validate proof signer
+     * @dev Separated for `stack too deep`
+     * @param didAddress - DID address, Used to revert
+     * @param proofSigner - Signer of proof
+     * @param info - Claim information
+     */
+    function _validateProofSigner(address didAddress, address proofSigner, ClaimInfo calldata info) internal virtual view {
+        bool isVerified;
+        uint index;
+        uint length = _trustedSigners.length();
+
+        while (index < length && !isVerified) {
+            if (proofSigner == _trustedSigners.at(index)) {
+                isVerified = true;
+            }
+            unchecked { ++index; }
+        }
+
+        if (!isVerified) {
+            revert InvalidProof(didAddress, info.typeId, info.xp, info.signature);
+        }
+    }
+
+    /**
+     * @notice Check the proofs and return the total amount of XP
+     * @param didAddress - DID address
+     * @param infos - Array of claim requests
+     * @return uint - Total amount of requested claim
+     */
+    function _validateClaimSignature(address didAddress, ClaimInfo[] calldata infos) internal virtual returns(uint) {
+        uint totalXP;
+
+        bytes memory rawMsg;
+        bytes32 dataHash;
+        address proofSigner;
+
+        uint infoLen = infos.length;
+        DateTime._DateTime memory curTime = DateTime.parseTimestamp(block.timestamp);
+
+        for (uint i; i < infoLen;) {
+            if (infos[i].xp == 0) {
+                revert InvalidXP(didAddress, infos[i].typeId, infos[i].xp);
+            }
+            _validateProofTime(didAddress, curTime, infos[i]);
+
+            rawMsg = abi.encodePacked(didAddress, infos[i].typeId, infos[i].issueYear, infos[i].issueMonth);
+            
+            // Check whether `typeId` is claimed in this month
+            if (isClaimedTypeId[rawMsg]) {
+                revert DuplicatedRequest(didAddress, infos[i].typeId, infos[i].xp, infos[i].signature);
+            }
+            isClaimedTypeId[rawMsg] = true;
+
+            if (bytes(infos[i].uniqueId).length != 0) {
```

Comment:
Please add more comments explaining the logic of this code.

I don't think the logic is correct. It doesn't appear to be verifying the `infos[i]` signature is valid based on the data provided.

Reply:
Third, it checks whether the `typeId` of the claim information is claimed in this month:
It check for the `${didAddress}${typeId}${year}${month}`.

Reply:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/4428b946-7e81-4725-b467-d51bf6d6f026)

Reply:
4th : if the `uniqueId` is optional, it checks whether this `uniqueId` is claimed before for the specified `typeId`:
![image](https://github.com/verida/blockchain-contracts/assets/15656252/2ba420d3-bc68-4d63-982e-02301d450453)

Reply:
5th: If the `uniqueId` is optional, it checks the `signature` is claimed before.
Here, the `signature` is encoded by the `did-typeid-year-month-xp`. So checking the `signature` means the checking the `did-typeId-year-mont-xp` value.
![image](https://github.com/verida/blockchain-contracts/assets/15656252/2360cd93-518d-457c-b4c7-7d193f246ab5)

Reply:
I was wanting your comments in the code itself.

As per my comment below, the `signature` isn't being verified correctly.

Reply:
Already explained about the Turing completness in the Solidity

Reply:
Added comments.
And the verification steps are updated.